examples/wmt/models.py

```diff
@@ -0,0 +1,774 @@
+# Lint as: python3
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer-based machine translation model."""
+
+from flax import nn
+from jax import lax
+import jax.numpy as jnp
+import numpy as np
+
+
+def shift_right(x, axis=1):
+  """Shift the input to the right by padding on axis 1."""
+  pad_widths = [(0, 0)] * len(x.shape)
+  pad_widths[axis] = (1, 0)
+  padded = jnp.pad(
+      x, pad_widths, mode='constant', constant_values=x.dtype.type(0))
+  return padded[:, :-1]
+
+
+def sinusoidal_init(max_len=2048,
+                    min_scale=1.0,
+                    max_scale=10000.0):
+  """1D Sinusoidal Position Embedding Initializer.
+
+  Args:
+      max_len: maximum possible length for the input
+      min_scale: float: minimum frequency-scale in sine grating.
+      max_scale: float: maximum frequency-scale in sine grating.
+
+  Returns:
+      output: init function returning `(1, max_len, d_feature)`
+  """
+
+  def init(key, shape, dtype=np.float32):
+    """Sinusoidal init."""
+    del key, dtype
+    d_feature = shape[-1]
+    pe = np.zeros((max_len, d_feature), dtype=np.float32)
+    position = np.arange(0, max_len)[:, np.newaxis]
+    scale_factor = -np.log(max_scale / min_scale) / (d_feature // 2 - 1)
+    div_term = min_scale * np.exp(np.arange(0, d_feature // 2) * scale_factor)
+    pe[:, :d_feature // 2] = np.sin(position * div_term)
+    pe[:, d_feature // 2: 2 * (d_feature // 2)] = np.cos(position * div_term)
+    pe = pe[np.newaxis, :, :]  # [1, max_len, d_feature]
+    return jnp.array(pe)
+
+  return init
+
+
+class AddPositionEmbs(nn.Module):
+  """Adds (optionally learned) positional embeddings to the inputs."""
+
+  def apply(self,
+            inputs,
+            inputs_positions=None,
+            max_len=512,
+            posemb_init=None,
+            cache=None):
+    """Applies AddPositionEmbs module.
+
+    By default this layer uses a fixed sinusoidal embedding table.  If a
+    learned position embedding is desired, pass an initializer to
+    posemb_init.
+
+    Args:
+      inputs: input data
+      inputs_positions: input position indices for packed sequences.
+      max_len: maximum possible length for the input
```

Comment:
No because it could be larger than that if you have multiple batch/length sizes in a dynamically padded dataset - so this is the maximum length across _all_ batched examples in the dataset.